File: benchmarks/medium/M2-csv-processing/python/main.py
import hashlib
import math
import os
import sys
import time

FIRST_NAMES = ["James", "Mary", "John", "Patricia", "Robert",
               "Jennifer", "Michael", "Linda", "David", "Elizabeth"]
LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Jones",
              "Garcia", "Miller", "Davis", "Rodriguez", "Martinez"]
DEPARTMENTS = ["Engineering", "Marketing", "Sales", "HR",
               "Finance", "Operations", "Support", "Legal"]

N = 1_000_000
RUNS = 10
CSV_FILE = "input.csv"
OUTPUT_FILE = "output.csv"


def generate_csv():
    state = 42
    with open(CSV_FILE, "w") as f:
        f.write("id,first_name,last_name,email,department,salary\n")
        for i in range(N):
            state = (state * 1664525 + 1013904223) & 0xFFFFFFFF
            salary = 30000.0 + (state / 4294967295.0) * 120000.0
            first = FIRST_NAMES[i % 10]
            last = LAST_NAMES[i % 10]
            dept = DEPARTMENTS[i % 8]
            f.write(f"{i + 1},{first},{last},{first.lower()}.{last.lower()}@company.com,{dept},{salary:.2f}\n")


def process():
    # Read CSV
    rows = []
    with open(CSV_FILE, "r") as f:
        f.readline()  # skip header
        for line in f:
            parts = line.rstrip('\n').split(',')
            salary = float(parts[5])
            if salary > 75000.0:
                rows.append((parts[4], salary))

    # Group by department
    groups = {}
    for dept, salary in rows:
        if dept not in groups:
            groups[dept] = [0, 0.0, -1e18, 1e18]
        g = groups[dept]
        g[0] += 1
        g[1] += salary
        if salary > g[2]:
            g[2] = salary
        if salary < g[3]:
            g[3] = salary

    # Build results
    results = []
    for dept, g in groups.items():
        count, total, mx, mn = g
        results.append((dept, count, total / count, mx, mn))

    # Sort by avg_salary descending
    results.sort(key=lambda x: -x[2])

    # Write output
    with open(OUTPUT_FILE, "w") as f:
        f.write("department,count,avg_salary,max_salary,min_salary\n")
        for dept, count, avg, mx, mn in results:
            f.write(f"{dept},{count},{avg:.2f},{mx:.2f},{mn:.2f}\n")


def main():
    timings = []

    print("=== BENCHMARK: M2 - CSV Processing ===")
    print("Language: Python")
    print(f"Runtime:  Python {sys.version.split()[0]}")
    print()

    generate_csv()

    for run in range(1, RUNS + 1):
        start = time.perf_counter()
        process()
        elapsed = (time.perf_counter() - start) * 1000
        timings.append(elapsed)
        print(f"Run {run:2d}: {elapsed:.2f} ms")

    print()
    min_t = min(timings)
    avg_t = sum(timings) / len(timings)
    max_t = max(timings)
    variance = sum((t - avg_t) ** 2 for t in timings) / len(timings)
    stddev = math.sqrt(variance)

    print(f"Min:    {min_t:.2f} ms")
    print(f"Avg:    {avg_t:.2f} ms")
    print(f"Max:    {max_t:.2f} ms")
    print(f"StdDev: {stddev:.2f} ms")
    print()

    # Verification: SHA-256 of output.csv
    h = hashlib.sha256()
    with open(OUTPUT_FILE, "rb") as f:
        h.update(f.read())
    print(f"Verification: SHA-256 = {h.hexdigest()}")

    os.remove(CSV_FILE)
    os.remove(OUTPUT_FILE)


if __name__ == "__main__":
    main()

File: benchmarks/medium/M2-csv-processing/javascript/main.js
const crypto = require('crypto');
const fs = require('fs');

const FIRST_NAMES = ["James", "Mary", "John", "Patricia", "Robert",
                     "Jennifer", "Michael", "Linda", "David", "Elizabeth"];
const LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Jones",
                    "Garcia", "Miller", "Davis", "Rodriguez", "Martinez"];
const DEPARTMENTS = ["Engineering", "Marketing", "Sales", "HR",
                     "Finance", "Operations", "Support", "Legal"];

const N = 1_000_000;
const RUNS = 10;
const CSV_FILE = "input.csv";
const OUTPUT_FILE = "output.csv";

function generateCsv() {
    const lines = ["id,first_name,last_name,email,department,salary"];
    let state = 42;
    for (let i = 0; i < N; i++) {
        state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
        const salary = 30000.0 + (state / 4294967295.0) * 120000.0;
        const first = FIRST_NAMES[i % 10];
        const last = LAST_NAMES[i % 10];
        const dept = DEPARTMENTS[i % 8];
        lines.push(`${i + 1},${first},${last},${first.toLowerCase()}.${last.toLowerCase()}@company.com,${dept},${salary.toFixed(2)}`);
    }
    fs.writeFileSync(CSV_FILE, lines.join('\n') + '\n');
}

function processCsv() {
    const data = fs.readFileSync(CSV_FILE, 'utf8');
    const lines = data.split('\n');

    // Read, filter, group in one pass
    const groups = new Map();
    // Skip header (index 0) and empty trailing line
    for (let i = 1; i < lines.length; i++) {
        const line = lines[i];
        if (line.length === 0) continue;
        const parts = line.split(',');
        const salary = parseFloat(parts[5]);
        if (salary > 75000.0) {
            const dept = parts[4];
            let g = groups.get(dept);
            if (g === undefined) {
                g = { count: 0, total: 0, max: -Infinity, min: Infinity };
                groups.set(dept, g);
            }
            g.count++;
            g.total += salary;
            if (salary > g.max) g.max = salary;
            if (salary < g.min) g.min = salary;
        }
    }

    // Build and sort results
    const results = [];
    for (const [dept, g] of groups) {
        results.push({ dept, count: g.count, avg: g.total / g.count, max: g.max, min: g.min });
    }
    results.sort((a, b) => b.avg - a.avg);

    // Write output
    const out = ["department,count,avg_salary,max_salary,min_salary"];
    for (const r of results) {
        out.push(`${r.dept},${r.count},${r.avg.toFixed(2)},${r.max.toFixed(2)},${r.min.toFixed(2)}`);
    }
    fs.writeFileSync(OUTPUT_FILE, out.join('\n') + '\n');
}

function main() {
    const timings = [];

    console.log("=== BENCHMARK: M2 - CSV Processing ===");
    console.log("Language: JavaScript");
    console.log(`Runtime:  Node.js ${process.version}`);
    console.log();

    generateCsv();

    for (let run = 1; run <= RUNS; run++) {
        const start = process.hrtime.bigint();
        processCsv();
        const elapsed = Number(process.hrtime.bigint() - start) / 1e6;
        timings.push(elapsed);
        console.log(`Run ${String(run).padStart(2)}: ${elapsed.toFixed(2)} ms`);
    }

    console.log();
    const min = Math.min(...timings);
    const max = Math.max(...timings);
    const avg = timings.reduce((a, b) => a + b, 0) / timings.length;
    const variance = timings.reduce((a, t) => a + (t - avg) ** 2, 0) / timings.length;
    const stddev = Math.sqrt(variance);

    console.log(`Min:    ${min.toFixed(2)} ms`);
    console.log(`Avg:    ${avg.toFixed(2)} ms`);
    console.log(`Max:    ${max.toFixed(2)} ms`);
    console.log(`StdDev: ${stddev.toFixed(2)} ms`);
    console.log();

    // Verification: SHA-256 of output.csv
    const hash = crypto.createHash('sha256').update(fs.readFileSync(OUTPUT_FILE)).digest('hex');
    console.log(`Verification: SHA-256 = ${hash}`);

    fs.unlinkSync(CSV_FILE);
    fs.unlinkSync(OUTPUT_FILE);
}

main();
